File: FieldKit.Modules.Visit/Geofencing.cs
using FieldKit.SharedKernel;

namespace FieldKit.Modules.Visit;

/// <summary>What a check-in's position means, before anything is written down.</summary>
/// <param name="DistanceMetres">
/// How far the rep was from the outlet, or null when that could not be worked out — see
/// <see cref="Geofencing.Assess"/>.
/// </param>
public sealed record GeofenceAssessment(bool Inside, double? DistanceMetres, bool ReasonRequired);

/// <summary>
/// Whether a rep checking in is at the outlet, and whether that needs explaining
/// (<c>VIS-01</c>, <c>VIS-02</c>, <c>BR-VIS-2</c>).
/// </summary>
/// <remarks>
/// <para>
/// <b>Pure</b>, like <c>PriceResolver</c> and <c>JourneyGenerator</c>: two positions, a radius and a
/// policy flag in, one assessment out. The same rules have to run on a device that is offline
/// (<c>§7</c>), so they cannot be entangled with a database — and a rule about where somebody is
/// standing is exactly the kind that is argued about, which means it has to be testable by writing
/// down two coordinates.
/// </para>
/// <para>
/// <b>Never blocks.</b> <c>BR-VIS-2</c> is emphatic: outside the geofence the visit is still
/// allowed, it just has to be explained. Nothing here returns "no" — the strongest thing it says is
/// <see cref="GeofenceAssessment.ReasonRequired"/>, and refusing a check-in that arrives without one
/// is the endpoint asking for the sentence, not the rule turning a rep away from a shop.
/// </para>
/// </remarks>
public static class Geofencing
{
    /// <summary>Mean Earth radius, in metres — the sphere the haversine formula assumes.</summary>
    private const double EarthRadiusMetres = 6_371_000;

    /// <summary>
    /// Assesses a check-in position.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Three things make a reason unnecessary, and they are different in kind:
    /// </para>
    /// <list type="bullet">
    /// <item><b>The rep is inside the radius.</b> Nothing to explain.</item>
    /// <item><b>The channel does not expect presence</b> (<c>IVisitWorkflow</c>). A phone call is
    /// legitimately not at the outlet, and demanding a reason would record an exception where
    /// nothing exceptional happened — the assumption <c>BR-VIS-2</c> spells out.</item>
    /// <item><b>The outlet has no coordinates.</b> Nobody has placed the shop, so "were you there?"
    /// has no answer — and making a rep justify a gap in master data would blame them for it. The
    /// visit records what it knows: a position, and no distance.</item>
    /// </list>
    /// <para>
    /// The device having no fix is the one case that <i>does</i> need explaining even though nothing
    /// can be measured: a rep whose phone reports no position at a shop that has one is exactly the
    /// case a supervisor would want to see, and it is also how a check-in would be faked.
    /// </para>
    /// </remarks>
    /// <param name="at">Where the rep says they are, or null when the device had no fix.</param>
    /// <param name="outlet">Where the outlet is, or null when nobody has placed it.</param>
    /// <param name="radiusMetres">How close counts as there.</param>
    /// <param name="presenceExpected">Whether being at the outlet is part of this visit at all.</param>
    public static GeofenceAssessment Assess(
        GeoPoint? at, GeoPoint? outlet, int radiusMetres, bool presenceExpected)
    {
        if (!presenceExpected) return new GeofenceAssessment(Inside: false, null, ReasonRequired: false);

        if (outlet is null) return new GeofenceAssessment(Inside: false, null, ReasonRequired: false);

        if (at is not { } here) return new GeofenceAssessment(Inside: false, null, ReasonRequired: true);

        var distance = DistanceMetres(here, outlet.Value);
        var inside = distance <= radiusMetres;

        return new GeofenceAssessment(inside, distance, ReasonRequired: !inside);
    }

    /// <summary>
    /// Great-circle distance between two points, in metres.
    /// </summary>
    /// <remarks>
    /// Haversine on a sphere. Wrong by up to about half a percent against the real, slightly squashed
    /// Earth — twenty-odd centimetres over the distances this compares, against a GPS fix that is
    /// routinely tens of metres out. Reaching for Vincenty here would be precision spent on the one
    /// term that is already the smallest source of error.
    /// </remarks>
    public static double DistanceMetres(GeoPoint from, GeoPoint to)
    {
        var lat1 = double.DegreesToRadians(from.Latitude);
        var lat2 = double.DegreesToRadians(to.Latitude);
        var deltaLat = lat2 - lat1;
        var deltaLon = double.DegreesToRadians(to.Longitude - from.Longitude);

        var a = (Math.Sin(deltaLat / 2) * Math.Sin(deltaLat / 2))
            + (Math.Cos(lat1) * Math.Cos(lat2) * Math.Sin(deltaLon / 2) * Math.Sin(deltaLon / 2));

        return 2 * EarthRadiusMetres * Math.Asin(Math.Min(1, Math.Sqrt(a)));
    }
}

File: frontend/lib/visits/geofencing.ts
/**
 * Whether a rep checking in is at the outlet (`VIS-01`, `VIS-02`, `BR-VIS-2`) — W9 slice 3.
 *
 * The device mirror of [`Geofencing`](../../../FieldKit.Modules.Visit/Geofencing.cs), and the third
 * rule this codebase implements twice. It has to run here because a rep standing in a shop with no
 * signal still has to be told whether they are inside the fence, and it has to *agree* with the
 * server because `IVisitIngest` stores the device's verdict **unmodified** — re-judging a visit
 * against today's radius would reclassify a rep who was legitimately inside yesterday's.
 *
 * <b>That makes disagreement worse here than in pricing.</b> A price recomputed server-side when an
 * order is placed gets a second opinion; this gets none. A device that decides "outside" writes a
 * supervisor an exception that never happened, permanently, and nothing downstream can tell.
 */

/** A position, as a device reports it or as an outlet records it. */
export type GeoPoint = { latitude: number; longitude: number };

/** What a check-in's position means, before anything is written down. */
export type GeofenceAssessment = {
  inside: boolean;
  /** How far the rep was, or `null` when that could not be worked out. */
  distanceMetres: number | null;
  reasonRequired: boolean;
};

/** Mean Earth radius, in metres — the sphere the haversine formula assumes. */
const EARTH_RADIUS_METRES = 6_371_000;

/**
 * Degrees to radians, spelled the way .NET spells it.
 *
 * `(degrees * PI) / 180`, **not** `degrees * (PI / 180)`. They are the same number in algebra and
 * not always the same `double`: the second folds `PI / 180` into a constant that is itself rounded,
 * then rounds again on the multiply. `double.DegreesToRadians` — which the C# original calls — is
 * the first form, so this is too.
 *
 * **Measured, because the temptation is to overstate it.** Across the 354 generated cases the two
 * forms give different answers on 105 of them, and the largest difference is **1.4 nanometres** —
 * three orders of magnitude below the tolerance the vectors compare with, and far too small to flip
 * a verdict given the generator's millimetre boundary guard. So this is not what keeps the two
 * languages agreeing; it is free, and it keeps them bit-identical where they can be, which makes any
 * *real* divergence stand out instead of hiding in noise that was there anyway.
 */
function toRadians(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

/**
 * Assesses a check-in position.
 *
 * Three things make a reason unnecessary, and they are different in kind:
 *
 * - **The rep is inside the radius.** Nothing to explain.
 * - **The channel does not expect presence.** A phone call is legitimately not at the outlet, and
 *   demanding a reason would record an exception where nothing exceptional happened.
 * - **The outlet has no coordinates.** Nobody has placed the shop, so "were you there?" has no
 *   answer — and making a rep justify a gap in master data would blame them for it.
 *
 * The device having no fix is the one case that *does* need explaining even though nothing can be
 * measured: a phone reporting no position at a shop that has one is exactly the case a supervisor
 * would want to see, and it is also how a check-in would be faked.
 *
 * **Nothing here blocks.** `BR-VIS-2` is emphatic — outside the fence the visit still happens, it
 * just has to be explained. The strongest thing this says is `reasonRequired`.
 */
export function assess(
  at: GeoPoint | null,
  outlet: GeoPoint | null,
  radiusMetres: number,
  presenceExpected: boolean,
): GeofenceAssessment {
  if (!presenceExpected) return { inside: false, distanceMetres: null, reasonRequired: false };

  if (outlet === null) return { inside: false, distanceMetres: null, reasonRequired: false };

  if (at === null) return { inside: false, distanceMetres: null, reasonRequired: true };

  const distance = distanceMetres(at, outlet);
  const inside = distance <= radiusMetres;

  return { inside, distanceMetres: distance, reasonRequired: !inside };
}

/**
 * Great-circle distance between two points, in metres.
 *
 * Haversine on a sphere, in the same order of operations as the C# original — which is deliberate
 * and is the only lever this file has over agreement. Every step is IEEE-754 double arithmetic, so
 * `+`, `*`, `/` and `Math.sqrt` are exactly reproducible; `Math.sin`, `Math.cos` and `Math.asin` are
 * **not**. Neither language's library is correctly rounded, and they are different libraries, so two
 * correct implementations may differ in the last bit or two.
 *
 * That is why the shared vectors compare distances with a tolerance rather than for equality, and
 * why the generator keeps its cases clear of the radius boundary — see `vectors/README.md`. Matching
 * the operation order costs nothing and keeps the disagreement in the last bits rather than letting
 * it compound.
 *
 * Wrong by up to about half a percent against the real, slightly squashed Earth — twenty-odd
 * centimetres over the distances this compares, against a GPS fix routinely tens of metres out.
 */
export function distanceMetres(from: GeoPoint, to: GeoPoint): number {
  const lat1 = toRadians(from.latitude);
  const lat2 = toRadians(to.latitude);
  const deltaLat = lat2 - lat1;
  const deltaLon = toRadians(to.longitude - from.longitude);

  const a =
    Math.sin(deltaLat / 2) * Math.sin(deltaLat / 2) +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(deltaLon / 2) * Math.sin(deltaLon / 2);

  return 2 * EARTH_RADIUS_METRES * Math.asin(Math.min(1, Math.sqrt(a)));
}
